<a href="https://colab.research.google.com/github/Vishal-3600/100-Days-of-ML/blob/main/100_Days_of_ML.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

.

🔵 **DAY 1 – NumPy (Foundation of ML)**
🎯 Why NumPy?

All ML libraries like:

TensorFlow

PyTorch

scikit-learn

are built on NumPy arrays.

NumPy = Fast matrix operations (C-level speed)

🔹 **1. Creating Arrays**

In [ ]:
import numpy as np

# 1D array
a=np.array([1,2,3,4])

#2D array
b=np.array([[1,2,3,],
          [5,6,7]])# in this 2D you must have same no of elements in its lower dimensions

c = np.array([[[1,2],[3,4]],
              [[5,6],[7,8]]])
print(a)
print(b)
print(c)

[1 2 3 4]
[[1 2 3]
 [5 6 7]]
[[[1 2]
  [3 4]]

 [[5 6]
  [7 8]]]


**2. Important Attributes**

In [ ]:
print(a.shape)# dimensions(meaning: 4 elements and 1 dimensions)
print(b.shape)#(Meaning 2 rows and 3 columns)
print(c.shape)#(meaning 2 block 2 rows and 2 columns) This is 3D array
'''
Layer 1
1 2
3 4

Layer 2
5 6
7 8

'''
print(a.ndim) #number of axes(dimenisons)
print(b.ndim)
print(c.ndim)
print(a.dtype)# dataType

(4,)
(2, 3)
(2, 2, 2)
1
2
3
int64


**3. Vectorized Operations (Very Important)**

In [ ]:
x = np.array([1,2,3])
y = np.array([4,5,6])

print(x + y)      # element-wise addition
print(x * y)      # element-wise multiplication
print()

#2D arrays
x = np.array([[1,2,3],
              [4,5,6]])

y = np.array([[10,20,30],
              [40,50,60]])


print(x + y)
print(x * y) #NumPy still does element-wise operations, but shapes must be compatible.
print()

#3D arrays
x = np.array([[[1,2],[3,4]],
              [[5,6],[7,8]]])

y = np.array([[[1,1],[1,1]],
              [[2,2],[2,2]]])

print(x + y)
print(x * y) #Still element-wise:(Same block(2), Same row(2), Same column(2))

#If shapes don’t match:
x = np.array([1,2,3])
y = np.array([[1,2,3],
              [4,5,6]])

print(x + y)
#This works because of Broadcasting.

#What is Broadcasting?(lean more about it)

#  NumPy automatically expands smaller array to match bigger one.

#x = [1 2 3]
# this becomes
'''[1 2 3]
    [1 2 3]'''

[5 7 9]
[ 4 10 18]

[[11 22 33]
 [44 55 66]]
[[ 10  40  90]
 [160 250 360]]

[[[ 2  3]
  [ 4  5]]

 [[ 7  8]
  [ 9 10]]]
[[[ 1  2]
  [ 3  4]]

 [[10 12]
  [14 16]]]


**🔹 4. Dot Product (Core ML Concept)**

Used in:

Linear Regression    ----> Y=Xw

Neural Networks

SVM

Logistic Regression

In [ ]:
x = np.array([1,2,3])
y = np.array([4,5,6])

dot = np.dot(x, y)
print(dot) #1×4+2×5+3×6=4+10+18=32
print()

#2D × 2D (Matrix Multiplication)

x = np.array([[1,2],
              [3,4]])

y = np.array([[5,6],
              [7,8]])

print(np.dot(x, y))#(Am×n​)⋅(Bn×p​)=(m×p)((2×2) · (2×2) → (2×2))
print()


#2D × 1D
x = np.array([[1,2,3],
              [4,5,6]])   # shape (2,3)

y = np.array([1,1,1])     # shape (3,)
print(np.dot(x, y))

32

[[19 22]
 [43 50]]

[ 6 15]


**5. Matrix Multiplication**

In [ ]:
#(2D × 2D)
A = np.array([[1,2],
              [3,4]])

B = np.array([[5,6],
              [7,8]])

print(np.matmul(A, B))

#3D × 3D (Batch Matrix Multiplication)
A = np.ones((2, 2, 3))   # shape (batch=2, 2, 3)
B = np.ones((2, 3, 4))   # shape (batch=2, 3, 4)

C = np.matmul(A, B)
print(C.shape)
print()
# Broadcasting in matmul
A = np.ones((2, 2, 3))   # (2,2,3)
B = np.ones((3, 4))      # (3,4) it will be treated as (1,3,4)

C = np.matmul(A, B) #(2,2,3) X (2,3,4) =(2,2,4)
print(C.shape)

# Rules
#Batch → Rows → Columns
#(last-2-dim) × (last-2-dim)
#All other dimensions are treated as batches.

[[19 22]
 [43 50]]
(2, 2, 4)

(2, 2, 4)


**@==Same as matmul()**

In [ ]:
# Mostly used in neural Networks
X = np.ones((32, 100))    # 32 samples
W = np.ones((100, 50))    # 50 neurons

#np.ones() is a NumPy function that creates an array filled entirely with 1s. All values are float by default --> to change into integer--> (np.ones((2,2), dtype=int))
print(np.ones((2,3))) # 2 rows and 3 columns

Y = X @ W
print(Y.shape)

[[1. 1. 1.]
 [1. 1. 1.]]
(32, 50)


**Day 1 Task**

**Create 3x3 random matrix.**

Compute:

1.Transpose

2.Determinant

3.Inverse

In [ ]:
A = np.array([[1,2],
              [3,4]])
print(np.transpose(A))
print(np.linalg.det(A))
print(np.linalg.inv(A))

[[1 3]
 [2 4]]
-2.0000000000000004
[[-2.   1. ]
 [ 1.5 -0.5]]


🔵 **DAY 2 – Pandas (Data Handling)**

 Why Pandas?

ML = 80% Data Cleaning

**🔹 1. Load Dataset**

In [8]:
import pandas as pd

df = pd.read_csv("/content/drive/MyDrive/Colab Notebooks/Titanic-Dataset (1).csv")
df.head()

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S


**🔹 2. Basic Info**

In [13]:
df.info() #Gives summary information about the DataFrame


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


**df.info** tells:

1.Number of rows

2.Column names

3.Data types(check whether all columns have same data type or not)

4.Non-null values

5.Memory usage

For Example:

**Age  714 non-null  float64**

Means:

Column name → Age

714 values are present

Remaining are missing

Data type = float

In [25]:
df.describe() #Gives statistical summary of numerical columns.

,PassengerId,Survived,Pclass,Age,SibSp,Parch,Fare
count,889.000000,889.000000,889.000000,889.000000,889.000000,889.000000,889.000000
mean,446.000000,0.382452,2.311586,29.653446,0.524184,0.382452,32.096681
std,256.998173,0.486260,0.834700,12.968366,1.103705,0.806761,49.697504
min,1.000000,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,224.000000,0.000000,2.000000,22.000000,0.000000,0.000000,7.895800
50%,446.000000,0.000000,3.000000,29.699118,0.000000,0.000000,14.454200
75%,668.000000,1.000000,3.000000,35.000000,1.000000,0.000000,31.000000
max,891.000000,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


**What Each Row Means**

Row	Meaning:

1. count:	Number of non-null values

2. mean:	Average

3. std:	Standard deviation

4. min:	Minimum value

5. 25%:	First quartile

6. 50%:	Median

7. 75%:	Third quartile

8. max:	Maximum value

In [26]:
df.describe(include='object') #Gives statistical summary of categorical columns.

,Name,Sex,Ticket,Cabin,Embarked
count,889,889,889,202,889
unique,889,2,680,146,3
top,"Dooley, Mr. Patrick",male,CA. 2343,C23 C25 C27,S
freq,1,577,7,4,644



**🔹 3. Handling Missing Values**

In [27]:
df.isnull().sum()

df['Age'].fillna(df['Age'].mean())
df.dropna(subset=['Embarked'])


,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.000000,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.000000,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.000000,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.000000,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.000000,0,0,373450,8.0500,NaN,S
...,...,...,...,...,...,...,...,...,...,...,...,...
886,887,0,2,"Montvila, Rev. Juozas",male,27.000000,0,0,211536,13.0000,NaN,S
887,888,1,1,"Graham, Miss. Margaret Edith",female,19.000000,0,0,112053,30.0000,B42,S
888,889,0,3,"Johnston, Miss. Catherine Helen ""Carrie""",female,29.699118,1,2,W./C. 6607,23.4500,NaN,S
889,890,1,1,"Behr, Mr. Karl Howell",male,26.000000,0,0,111369,30.0000,C148,C


🔹 **df.isnull()**

Returns a DataFrame of True / False

True=missing value

**.sum()**

.sum() counts number of True values. for each value in column form

**df['Age'].fillna(df['Age'].mean())**

Replaces missing values in the Age column with the mean (average) age.

**Why Use Mean?**

For numerical columns, common strategies:

Mean-->	Normal distribution

Median-->	Skewed data

Mode-->	Categorical data

**df.dropna(subset=['Embarked'])**

Removes rows where Embarked(पोतारोहण करना) column has missing value.

1.df.dropna() 👉 axis=0 → drop rows

2.df.dropna(axis=1) 👉 axis=1 → drop columns

**Drop if very few rows**

🔹 **4. Filtering Data**

Boolean Indexing (Conditional Filtering) in Pandas

Filtering is used for:

Exploratory Data Analysis (EDA)

Subgroup analysis

Data cleaning

Feature engineering

In [31]:
df[df['Sex'] == 'male']  # select all rows that has sex as male
df[df['Age'] > 30]
df.loc[df['Age'] > 30, 'Age'] = 50 # if you want to update the age based on the given condition
# both above do not change the original dataFrame they create a new DataFrame

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
11,12,1,1,"Bonnell, Miss. Elizabeth",female,58.0,0,0,113783,26.5500,C103,S
...,...,...,...,...,...,...,...,...,...,...,...,...
873,874,0,3,"Vander Cruyssen, Mr. Victor",male,47.0,0,0,345765,9.0000,NaN,S
879,880,1,1,"Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)",female,56.0,0,1,11767,83.1583,C50,C
881,882,0,3,"Markun, Mr. Johann",male,33.0,0,0,349257,7.8958,NaN,S
885,886,0,3,"Rice, Mrs. William (Margaret Norton)",female,39.0,0,5,382652,29.1250,NaN,Q


5. **GroupBy (Very Important)**


In [32]:
df.groupby("Sex")["Survived"].mean()

,Survived
Sex,
female,0.740385
male,0.188908
